### PR TITLE
bitfield encoder

### PR DESCRIPTION
**Pull request by laktech**
This pr includes the abaility to write iabtcf compliant byte arrays using BitWriter and encoding bit
fields and ranges.

### CODE REVIEW

File: iabtcf-encoder/src/main/java/com/iabtcf/encoder/BitFieldEncoder.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+            // emit range bits
+            bv.write(true, FieldDefs.IS_A_RANGE);
+
+            bv.write(numEntries, FieldDefs.NUM_ENTRIES);
+            bv.write(rangeBits);
+        } else {
+            // emit bit field
+            bv.write(false, FieldDefs.IS_A_RANGE);
+
+            byte[] bits = vendors.toByteArray();
```
Comment (chrispaterson) [inline iabtcf-encoder/src/main/java/com/iabtcf/encoder/BitFieldEncoder.java]: I don't understand why you need all this `byte` stuff???  I would think that you would just do:

```Java
for(int i = 0; i < maxVendorId; i++) {
  bv.write(vendors.get(i));
}
```

File: iabtcf-encoder/src/main/java/com/iabtcf/encoder/BitFieldEncoder.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    public BitWriter build() {
+        BitWriter bv = new BitWriter();
+
+        if (vendors.length() == 0) {
+            bv.write(0, FieldDefs.CORE_VENDOR_MAX_VENDOR_ID);
+            bv.write(false, FieldDefs.CORE_VENDOR_IS_RANGE_ENCODING);
+            return bv;
+        }
+
+        maxVendorId = maxVendorId == -1 ? vendors.length() : maxVendorId;
```
[the remark this answers is not in the document]
Comment (laktech) [reply iabtcf-encoder/src/main/java/com/iabtcf/encoder/BitFieldEncoder.java]: i agree, we should remove this.

File: iabtcf-encoder/src/main/java/com/iabtcf/encoder/BitFieldEncoder.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    }
+
+    /**
+     * Returns a BitWriter that represents the encoded vendor ids. The BitWriter may either encode a
+     * bit field or a range encoding; depending on which is smaller.
+     */
+    public BitWriter build() {
+        BitWriter bv = new BitWriter();
+
+        if (vendors.length() == 0) {
```
[the remark this answers is not in the document]
Comment (laktech) [reply iabtcf-encoder/src/main/java/com/iabtcf/encoder/BitFieldEncoder.java]: yes, this makes sense

File: iabtcf-encoder/src/main/java/com/iabtcf/encoder/BitFieldEncoder.java [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+                rangeBits.write(idxSet + 1, FieldDefs.START_OR_ONLY_VENDOR_ID);
+                rangeBits.write(length, FieldDefs.END_VENDOR_ID);
+            }
+            numEntries++;
+        } while ((idxSet = vendors.nextSetBit(idxClr)) > 0 && rangeBits.length() < vendors.length());
+
+        // emit max vendor id
+        bv.write(maxVendorId, FieldDefs.CORE_VENDOR_MAX_VENDOR_ID);
+
+        if (rangeBits.length() < vendors.length()) {
```
[the remark this answers is not in the document]
Comment (laktech) [reply iabtcf-encoder/src/main/java/com/iabtcf/encoder/BitFieldEncoder.java]: yeah, nice catch.  i'll address this

File: iabtcf-encoder/src/main/java/com/iabtcf/encoder/BitFieldEncoder.java (hunk 1 above)
Comment (laktech) [reply iabtcf-encoder/src/main/java/com/iabtcf/encoder/BitFieldEncoder.java]: yeah we can do that

File: iabtcf-encoder/src/main/java/com/iabtcf/encoder/BitFieldEncoder.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (laktech) [reply iabtcf-encoder/src/main/java/com/iabtcf/encoder/BitFieldEncoder.java]: actually, we do need this check due to the do-while loop

File: iabtcf-encoder/src/main/java/com/iabtcf/encoder/BitFieldEncoder.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (laktech) [reply iabtcf-encoder/src/main/java/com/iabtcf/encoder/BitFieldEncoder.java]: i think it's fine. so when the loop above stops when either 1) we've encoded all the vendors or 2) rangeBits.length >= vendors.length (not all vendor ids were encoded)

if (1) is true, then range encoding is using less bits and the if statement on line 128 should be true and we emit the range encoding

if (2) is true, then rangeBits.length() < vendors.length() is false.

the key is that the while loop **continues** as long as rangeBits.length is less than vendors.length. Hence, we only break out of the loop when that condition on line 128 is false (and assuming there are still more vendors to be written).

File: iabtcf-encoder/src/main/java/com/iabtcf/encoder/BitFieldEncoder.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+            add(i.nextInt());
+        }
+
+        return this;
+    }
+
+    /**
+     * Emit the specified max vendor id. By default, the maximum vendorId that was added is emitted.
+     */
+    public BitFieldEncoder setMaxVendorId(int maxVendorId) {
```
[the remark this answers is not in the document]
Comment (laktech) [reply iabtcf-encoder/src/main/java/com/iabtcf/encoder/BitFieldEncoder.java]: we can remove this.. was nice for testing though! maybe i'll make it protected / private and mark it @VisibileForTesting?

File: iabtcf-encoder/src/main/java/com/iabtcf/encoder/BitFieldEncoder.java [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+        this(prototype.vendors.length() == 0 ? new BitSet() : prototype.vendors.get(0, prototype.vendors.length()),
+                prototype.maxVendorId);
+    }
+
+    private BitFieldEncoder(BitSet vendors, int maxVendorId) {
+        this.vendors = vendors;
+        this.maxVendorId = maxVendorId;
+    }
+
+    public BitFieldEncoder add(int vendorId) {
```
[the remark this answers is not in the document]
Comment (laktech) [reply iabtcf-encoder/src/main/java/com/iabtcf/encoder/BitFieldEncoder.java]: yep that's fine

**Pull request merged**